### PR TITLE
how to spend one day not coding: write docstrings!

### LINKED ISSUE #329: API documentation
thanks to @kylebarron we now have automatic API documentation plublished https://cogeotiff.github.io/rio-tiler/api/rio_tiler/colormap/ (ref #287) 

This is just a start and I'd love to make the doc better looking:
- allow examples to be displayed (I tried to change the pdocs template but failed)
- the Class and Function font are not bold enough (IMO)
- the whole docs seems a bit noisy sometime (mostly because of MRO 🤷‍♂️) 

I'm going to start a PR to change all the docstrings to a better style as I did in https://github.com/cogeotiff/rio-tiler/blob/c6f0e40769e38cbfa0921516243fb7460adbf613/rio_tiler/colormap.py#L68-L83

### PR DESCRIPTION
**Pull request by vincentsarago**
closes #329

### CODE REVIEW

File: rio_tiler/colormap.py [hunk 1]
```diff
@@ -138,11 +137,13 @@ class ColorMaps:
...
     Args:
-        data (dict): colormaps.
+        data (dict): colormaps. Defaults to `rio_tiler.colormap.DEFAULTS_CMAPS`.
 
     """
 
-    data: Dict[str, Union[str, numpy.array]] = attr.ib()
+    data: Dict[str, Union[str, numpy.array]] = attr.ib(
+        default=attr.Factory(lambda: DEFAULTS_CMAPS)
+    )
```
Comment (vincentsarago) [inline rio_tiler/colormap.py:146]: thanks @geospatial-jeff for this ;-)

Comment (vincentsarago) [reply rio_tiler/colormap.py:146]: Tag: non-doc change but I couldn't resist

Comment (kylebarron) [reply rio_tiler/colormap.py:146]: I'd suggest `DEFAULT_CMAPS` and `DEFAULT_CMAPS_FILES`

File: rio_tiler/io/cogeo.py [hunk 2]
```diff
@@ -617,47 +531,69 @@ def feature(
...
+            assert cog.src_dataset
+
+        >>> with rasterio.open(src_path) as src_dst:
+                with COGReader(None, src_dataset=src_dst) as cog:
+                    cog.tile(...)
 
     """
 
+    filepath: str = attr.ib()
+    src_dataset: Union[DatasetReader, DatasetWriter, MemoryFile, WarpedVRT] = attr.ib(
```
Comment (vincentsarago) [inline rio_tiler/io/cogeo.py:561]: surface src_dataset to allow the user to use an opened dataset

Comment (vincentsarago) [reply rio_tiler/io/cogeo.py:561]: Tag: non-doc change but I couldn't resist

File: rio_tiler/io/stac.py [hunk 3]
```diff
@@ -96,7 +98,7 @@ def _get_assets(
         yield asset
 
 
-def to_pystac_item(item: Union[None, Dict, pystac.Item]) -> Union[None, pystac.Item]:
+def _to_pystac_item(item: Union[None, Dict, pystac.Item]) -> Union[None, pystac.Item]:
```
Comment (vincentsarago) [inline rio_tiler/io/stac.py:101]: breaking change but this function doesn't have to be public

File: rio_tiler/models.py [hunk 4]
```diff
@@ -217,11 +232,26 @@ def post_process(
                 data = scale_dtype(ops(to_math_type(data)), numpy.uint8)
 
         return ImageData(
-            data, mask, crs=self.crs, bounds=self.bounds, assets=self.assets
+            data,
+            mask,
+            crs=self.crs,
+            bounds=self.bounds,
+            assets=self.assets,
+            metadata=self.metadata,
```
Comment (vincentsarago) [inline rio_tiler/models.py:240]: metadata was missing

File: rio_tiler/profiles.py [hunk 5]
```diff
@@ -52,12 +52,12 @@ def __init__(self):
     def get(self, key, default=None):
         """Like normal item access but return a copy of the key."""
         if key in (self.keys()):
-            return self.data[key].copy()
+            return deepcopy(self.data[key])
```
Comment (vincentsarago) [inline rio_tiler/profiles.py:55]: gives the same results (for tests) but I think it's cleaner/safer

File: rio_tiler/utils.py [hunk 6]
```diff
@@ -310,56 +275,47 @@ def _requested_tile_aligned_with_internal_tile(
 
 
 def render(
-    tile: numpy.ndarray,
+    data: numpy.ndarray,
```
Comment (vincentsarago) [inline rio_tiler/utils.py:278]: Breaking change!

File: tests/test_io_cogeo.py [hunk 7]
```diff
@@ -363,6 +363,37 @@ def test_cog_with_internal_gcps():
...
+            b1_stats = cog.stats()["1"]
+            assert b1_stats.max == 623
+
+            tile_z = 8
+            tile_x = 183
+            tile_y = 120
+            data, mask = cog.tile(tile_x, tile_y, tile_z)
+            assert data.shape == (1, 256, 256)
+            assert mask.shape == (256, 256)
+        # https://github.com/mapbox/rasterio/issues/2092
```
Comment (vincentsarago) [inline tests/test_io_cogeo.py:392]: fixed in rasterio 1.2.0

File: rio_tiler/io/base.py [hunk 8]
```diff
@@ -156,39 +260,89 @@ async def metadata(
...
 class MultiBaseReader(BaseReader, metaclass=abc.ABCMeta):
-    """MultiBaseReader Reader."""
+    """MultiBaseReader Reader.
+
+    This Reader is suited for dataset that are composed of multiple assets (e.g. STAC).
+
+    Args:
+        reader (rio_tiler.io.BaseReader): reader.
+        reader_options (dict, option): options to forward to the reader. Defaults to {}.
+        tms (morecantile.TileMatrixSet, optional): TileMatrixSet grid definition. Defaults to WebMercatorQuad.
```
Comment (kylebarron) [inline rio_tiler/io/base.py:342]:
```suggestion
        tms (morecantile.TileMatrixSet, optional): TileMatrixSet grid definition. Defaults to `WebMercatorQuad`.
```

File: rio_tiler/io/base.py [hunk 9]
```diff
@@ -210,7 +364,15 @@ def parse_expression(self, expression: str) -> Tuple:
     def info(  # type: ignore
         self, assets: Union[Sequence[str], str] = None, *args, **kwargs: Any
     ) -> Dict[str, Info]:
-        """Return metadata from multiple assets"""
+        """Return metadata from multiple assets.
+
+        Args:
+            assets (sequence of str or str, optional): assets to fetch info from. Requiered keyword argument.
```
Comment (kylebarron) [inline rio_tiler/io/base.py:370]:
```suggestion
            assets (sequence of str or str, optional): assets to fetch info from. Required keyword argument.
```

File: rio_tiler/io/base.py [hunk 10]
```diff
@@ -231,7 +393,18 @@ def stats(  # type: ignore
         assets: Union[Sequence[str], str] = None,
         **kwargs: Any,
     ) -> Dict[str, Dict[str, ImageStatistics]]:
-        """Return array statistics from multiple assets"""
+        """Return array statistics from multiple assets.
+
+        Args:
+            pmin (float, optional): Histogram minimum cut. Defaults to 2.0.
+            pmax (float, optional): Histogram maximum cut. Defaults to 98.0.
+            assets (sequence of str or str): assets to fetch info from. Requiered keyword argument.
```
Comment (kylebarron) [inline rio_tiler/io/base.py:401]: `Requiered` do a global search and replace

File: rio_tiler/io/stac.py [hunk 11]
```diff
@@ -118,15 +120,15 @@ class STACReader(MultiBaseReader):
...
-        minzoom (Optional[int]): Set minzoom for the tiles. Defaults to tms.minzoom.
-        minzoom (Optional[int]): Set maxzoom for the tiles. Defaults to tms.maxzoom.
-        include (Optional[Set[str]]): Only Include specific assets. Defaults to None.
-        exclude (Optional[Set[str]]): Exclude specific assets. Defaults to None.
-        include_asset_types (Optional[Set[str]]): Only include some assets base on their type. Defaults to None.
-        exclude_asset_types (Optional[Set[str]]): Exclude some assets base on their type. Defaults to None.
-        reader (BaseReader): rio-tiler Reader (default is set to rio_tiler.io.COGReader).
-        reader_options (dict): additional option to forward to the Reader (default is {}).
+        item (dict or pystac.Item, STAC): Stac Item.
+        minzoom (int, optional): Set minzoom for the tiles. Defaults to tms.minzoom.
```
Comment (kylebarron) [inline rio_tiler/io/stac.py:124]: I kinda like the `Optional[int]` syntax, but ok

File: rio_tiler/io/cogeo.py [hunk 12]
```diff
@@ -26,63 +26,37 @@
...
B1/B20", max_size=1024)
-        Read part of the COG.
-    preview(max_size=1024)
-        Read preview of the COG.
-    point((10, 10), indexes=1)
-        Read a point value from the COG.
-    stats(pmin=5, pmax=95)
-        Get Raster statistics.
-    meta(pmin=5, pmax=95)
-        Get info + raster statistics
```
Comment (kylebarron) [inline rio_tiler/io/cogeo.py:85]: Properties and methods are just removed?

File: rio_tiler/profiles.py (hunk 5 above)
Comment (kylebarron) [reply rio_tiler/profiles.py:55]: It depends if `self.data[key]` ever has deeply-nested objects.

**Review by kylebarron — APPROVED** [review]

File: rio_tiler/io/base.py (hunk 8 above)
Comment (kylebarron) [reply rio_tiler/io/base.py:342]: Eh you don't need to do merge this because it won't be widespread.

File: rio_tiler/io/cogeo.py (hunk 12 above)
Comment (vincentsarago) [reply rio_tiler/io/cogeo.py:85]: They should now be rendered directly by pdocs

File: rio_tiler/profiles.py (hunk 5 above)
Comment (vincentsarago) [reply rio_tiler/profiles.py:55]: I don't think so 🤷‍♂️ I guess we can revert

File: rio_tiler/io/stac.py (hunk 11 above)
Comment (vincentsarago) [reply rio_tiler/io/stac.py:124]: seems that using `type` notation is not yet standard. we can update this later

**Pull request merged**